src/Shared/FileMatcher.cs

```diff
@@ -1012,10 +1032,27 @@ struct RecursionState
             {
                 return Enumerable.Empty<string>();
             }
+
+            // Back-compat hack: We don't use case-insensitive file enumeration I/O on Linux so the behavior is different depending
+            // on the NeedsToProcessEachFile flag. If the flag is false and matching is done within the _getFileSystemEntries call,
+            // it is case sensitive. If the flag is true and matching is handled with MatchFileRecursionStep, it is case-insensitive.
+            // TODO: Can we fix this by using case-insensitive file I/O on Linux?
+            bool forceFileProcessing = false;
+            string filespec;
+            if (NativeMethodsShared.IsLinux && recursionState.SearchData.DirectoryPattern != null)
+            {
+                filespec = "*.*";
+                forceFileProcessing = true;
+            }
+            else
+            {
+                filespec = recursionState.SearchData.Filespec;
+            }
```

Comment:
@cdmihai may I please get your opinion on this? Glob matching on Linux is currently sometimes case-sensitive and sometimes case-insensitive depending on which code path we happen to take in this function.

In this PR I am optimizing globbing and without the hack/workaround a couple of test cases fail on Linux because they expect certain behavior w.r.t. case sensitivity. These assumptions are encoded very explicitly:

https://github.com/dotnet/msbuild/blob/6bad8ea6a27fc23621ee8a41fd941dff01976de1/src/Shared/UnitTests/FileMatcher_Tests.cs#L112-L120

and for example:

https://github.com/dotnet/msbuild/blob/6bad8ea6a27fc23621ee8a41fd941dff01976de1/src/Shared/UnitTests/FileMatcher_Tests.cs#L219

How would you feel about fixing Linux to be consistent, so most likely case-insensitive like other platforms, by changing file enumeration calls to pass [EnumerationOptions ](https://docs.microsoft.com/en-us/dotnet/api/system.io.enumerationoptions?view=net-5.0) with `MatchCasing=false`? Technically a breaking change, yes, but the current state makes it hard to innovate and has us write ugly hacks like this one.

Reply:
Not completely relevant, but a bit of a discussion happened on linux case sensitivity here: https://github.com/dotnet/msbuild/issues/5658

Reply:
Here's a simple example where we are inconsistent today.

Source tree:
```
.
├── abc
│   ├── File
│   └── file
└── repro.proj
```

repro.proj:
```xml
<Project>
  <ItemGroup>
    <Item1 Include="a*/f*" />
    <Item2 Include="abc/f*" />
  </ItemGroup>

  <Target Name="Build">
    <Message Text="@(Item1)" />
    <Message Text="@(Item2)" />
  </Target>
</Project>
```

```sh
$ dotnet build repro.proj /v:n
```

Prints:
```
  abc/File;abc/file
  abc/file
```

i.e. the first glob matched both ```file``` and ```File``` while the second one did not.

Reply:
Case sensitivity is part of a long standing undesigned area of MSBuild: path representation. See #781 and #1024. It's pretty much a can of worms.

However way we change it, we'll probably break *somebody* and [as noted before](https://github.com/dotnet/msbuild/issues/1024#issuecomment-549564420), it's probably of minimal user benefit.
That being said, it would be interesting to test these assumptions and hardcode casing to case insensitive, provided that an escape hatch to old behaviour is provided. Unfortunately the escape hatch would keep the messy code, but we could at least give it a year or so to hear any complaints before deleting it.

If you do choose to embark on this journey, I'd prefer the following:
- design it and take the other aspects of #781 and #1024 into account, even if just as placeholders for future design. The make it an official spec on docs.microsoft.com on how msbuild handles file system configurations.
- make all the msbuild code base uniform to the new spec, not just globbing
- have a central place in the codebase that decides file system configuration policies (similar to how [BuildEnvironment](https://github.com/dotnet/msbuild/blob/f0eebf2872d76ab0cd43fdc4153ba636232b222f/src/Shared/BuildEnvironmentHelper.cs#L497) centralizes msbuild file and assembly structure) instead of having it duplicated all over the place. One policy would be the `Legacy` policy, which is the behaviour today. Hopefully we can delete it in 1-2 years if nothing really critical breaks.

Reply:
Thank you for the details. It sounds like changing the behavior should be done separately so I am keeping the Linux workaround and this PR overall strives for 100% back compat.